source/Calamari/Behaviors/TargetDiscoveryBehaviour.cs

```diff
@@ -153,10 +154,25 @@ private IEnumerable<IDeploymentSlot> ListDeploymentSlots(IWebApp webApp)
         {
             var policy = CreateAzureQueryRetryPolicy(5, $"listing deployment slots for web app {webApp.Name}");
 
-            return policy.Execute(() =>
+            try
             {
-                return webApp.DeploymentSlots.List();
-            });
+                return policy.Execute(() =>
+                {
+                    return webApp.DeploymentSlots.List();
+                });
+            }
+            catch (DefaultErrorResponseException dex)
+            {
+                if (dex.Response.StatusCode == System.Net.HttpStatusCode.NotFound)
+                {
+                    // We could get a 404 listing slots if the web app gets deleted 
+                    // after being found but before we can check it's slots, in this
+                    // case we'll log a message and continue
+                    Log.Verbose($"Could not list deployment slots for web app {webApp.Name} as it could no longer be found");
+                    return Enumerable.Empty<IDeploymentSlot>();
+                }
+                throw;
+            }
```

Comment:
I found it difficult to understand the interplay between the outer `try/catch` and the Policy execution. I got there, but it was a bit of mental load.

Another way we could express this a little more cleanly could be using the [Fallback](https://github.com/App-vNext/Polly#fallback) policy combined with a [PolicyWrap](https://github.com/App-vNext/Polly/wiki/PolicyWrap). 

A brief example, for which I have a slightly larger Linqpad script demonstrating how it works, if you're interested:

```c#
var rateLimitingPolicy = CreateAzureQueryRetryPolicy(5, $"listing deployment slots for web app {webApp.Name}");

var fallbackPolicy = Policy<IEnumerable<DeploymentSlot>>
    .Handle<DefaultErrorResponseException>(dex => dex.Response.StatusCode == System.Net.HttpStatusCode.NotFound;)
    .Fallback<IEnumerable<DeploymentSlot>>(
        fallbackValue: Enumerable.Empty<DeploymentSlot>(), 
        onFallback: (exception, context) => Log.Verbose($"Could not list deployment slots for web app {webApp.Name} as it could no longer be found")
);

return fallbackPolicy.Wrap(rateLimitingPolicy).Execute(() => 
{
    return webApp.DeploymentSlots.List();
}
```

Reply:
Having the fallback policy wrapping the rateLimitingPolicy means that if the conditions for the fallback are met, it gets executed and returns before trying the rateLimitingPolicy. This could also perhaps remove the need for the condition of `ex => ex.Response.StatusCode != System.Net.HttpStatusCode.NotFound` on the rateLimitingPolicy.

Reply:
Thanks, I've not used Fallback policies before, it does keep all the logic together a bit nicer rather than being split across policy and exception handling. I'll take a look at putting it in.

Reply:
(You might call it `webAppNotFoundPolicy` or something else more descriptive than than `fallbackPolicy` :))